docs/modules/jenkins-shared-library/partials/odsComponentStageScanWithAqua.adoc.tmpl

```diff
@@ -0,0 +1,43 @@
+The "Aqua Security Scan" stage scans an image that was previously build in that same pipeline run.
+
+As a result, a BitBucket Code Insight entry is added to the git commit (in BitBucket) that basically
+contains a link to the scan result on the Aqua platform. The BitBucket Code Insight entry can be seen in a Pull Request.
+A Pull Request in BitBucket shows the Code Insight of the latest commit of the PR.
+
+To get started, make sure you have a `ConfigMap` in your project-cd namespace in OpenShift that has these fields:
+----
+...
+metadata:
+  name: aqua
+...
+data:
+  registry: <registry-name-in-aqua-platform>
+  secretName: <secret-name-of-aqua-user-credentials>
+  url: <aqua-platform-url>
+----
+
+. `registry`: Refers to a name for the image registry given in the aqua platform by an aqua platform admin.
+. `secretName`: Name of the `Secret` that contains the credentials of the aqua platform user that is used for executing the scan command. That user needs to have scanner rights.
+. `url`: Base URL of the aqua platform.
+
+Now, to use the stage, add it in your `Jenkinsfile`, e.g. like this:
+----
+) { context ->
+    ...
+    odsComponentStageScanWithAqua(context)
```

Comment:
Maybe good to show that this should be after the build stage, and if there is this `orElse` thing, it should be inside?

Reply:
@michaelsauter @renedupont as mentioned above, I am not confident that scanning with Aqua should be a stage in the component's Jenkinsfile. Why? It's too easy to overcome. Rene, can you share your thoughts on my feedback above?

Reply:
I think it has to be in the component pipeline if you want feedback on pull requests. If you additionally want scanning to happen in the orchestration pipeline, it would need to be implemented in addition I think, not instead of the component pipeline scanning. Could be done, but maybe does not need to be done in this PR?

Reply:
Absolutely, scanning must happen in the component pipeline. However, I would like to consider security as a feature of the platform, and not as a feature of a component. In other words, I would not make Aqua scanning implemented in a dedicated function within the pipeline, but as an add-on to the build stage: build -> scanWithAqua. The latter should only be invoked though if Aqua is enabled for the current installation. Ideally, Aqua can be enabled/disabled by a cluster admin and does not depend on a config change and ensuing re-installation of ODS. Possible reasons are: future license expiry, breaking changes in Aqua, etc. which shall not block all builds. This was part of a previous discussion with Rene.

Reply:
I was under the assumption that your consideration to run that stage mandatorily was only meant for projects that use the orchestration pipeline, and not a general thing for every user of ODS. That is why I did not see it as part of this PR. I think when enforcing this stage to run in every build is something you want to add, this dedicated function can be used in the orchestration pipeline to do so.

Reply:
Hmm.. no, I was not seeing it this way. Aqua scanning is the implementation of a long standing security requirement which is definitely not optional for developers, but an integral part of building a safe application platform. Conceptually, it shall only be disable-able at the cluster level.

Reply:
hi all, sorry to jump into, but I felt there is a misunderstanding on what the expectations would be, so I am listing from my point of view what I think we should have/distinguish:

- ODS/EDP component pipeline (team SLC): to have a stage on the CICD that clearly can tell you/teams how secure/healthy the component is (so to get feedback from Aqua: make use of the reports, and then if needed and possible even get into the reports on the Aqua side to dig more) and depending on the phase of the project work over it as deeply as needed (we know that not always one will really need to get blocked at this point -> MVP's, explorations, tests branches, demos,...) and even depending on the criticallity of the project auto-enforce actions (like failing the pipeline) 

- Platform/Cluster: to ensure corporate policies and governance requirements, such as avoiding deployments of images that are not compliant (if the project is on PROD maybe only...), enforcing by blocking or auditing non-permitted processes/behaviours...depending on the type of project, criticallity, processes,...

Reply:
I don't think we deviate much. Ever since I've been involved in this, scans when building have never been considered optional though. Where do you see the issue with Aqua scanning every build?

Reply:
@metmajer So you are envisioning the Aqua scan as part of the build stage, not as a separate stage? I like that.

How to turn Aqua on/off at cluster level is less clear to me. So far we do something similar with the SonarQube edition by writing information into the Jenkins image. Therefore, to change the setting, a new image needs to be built AND rolled out to all users. It would be nice to flip a switch centrally, but I think right now we do not have such a thing. It would need to be a service which can be asked by the pipeline whether Aqua scanning should be performed. The only "central" service we have under control is the provisioning app, and I think that's probably the wrong place. Next to "central" information, the "local" information whether Aqua scanning should be on/off could be presence of the `aqua` ConfigMap (and/or a key within that).

@metmajer Did you have something specific in mind how that central information about Aqua can be exposed to the pipelines?

Reply:
An `aqua` ConfigMap seems like a good approach to me. Ideally, if it's located centrally in the `OpenDevStack` namespace. We might be out of luck with our current clusters though: [Is there a way to share a configMap in kubernetes between namespaces?
](https://stackoverflow.com/questions/55515594/is-there-a-way-to-share-a-configmap-in-kubernetes-between-namespaces/55516643).

Reply:
Ah right, I somehow thought we can't do that - guess I was thinking about services. Duh.

Accessing the content of a central `ConfigMap` should actually work with `oc`, provided that the `jenkins` serviceaccount has `view` permissions in the central namespace. And given https://github.com/opendevstack/ods-core/blob/master/ods-setup/setup-ods-project.sh#L56, this should be possible. I checked in a Jenkins pod on both OpenShift 4 and 3.11, and it works in both.

So in theory we can have a central `ConfigMap` acting as a switch for Aqua.

Reply:
Currently, the `aqua` ConfigMap is meant to be in the project cd namespace because it also contains the `secretName` of the aqua user credentials that are different for each project/team. If there is a better way of using it we can change.

Reply:
I am not convinced that enabling/disabling Aqua at a project level will increase the trust in the EDP/ODS Platform, especially if this can happen at the will of every project lead. A power switch at the cluster level is still beneficial.

Reply:
@renedupont The `url` would not be project specific though. Is `registry` a project specific value? Not entirely clear to me. 

The `secretName` should probably be something that can be changed on project level. What do you think of a layered approach? As in, `url`, `registry` and `secretName` can be set in a global config map, and the local config map can override those? The global config map would also contain a switch if Aqua is on or off (and this would not be overridable as far as I understand?). Does this sound good or too complicated?

Reply:
Hello @metmajer, considering that
- you already have a pretty good and precise vision on how to further shape this
- I understood that it soon needs to be also tested in your environment
- I will not have much time to spent on this topic for at least the next 2-3 sprints (~4-6wks)
I would suggest that you (or your team) takes it on from here to not cause any delays on finishing it the way you envision.
I pushed it already to a branch on the opendevstack repo: https://github.com/opendevstack/ods-jenkins-shared-library/tree/feature/aqua-stage-master
Please let me know if this works for you?

Reply:
@renedupont I would love to, but my team is blocked with other duties needed for the upcoming release as well. We were informed that you want to own the integration with Aqua. In this case, please inform security about your situation.

Reply:
@renedupont forgot to add: Then, let's see how we can continue and how your sprints go. Maybe an opportunity opens up our end, but I don't see that in the coming weeks.

Reply:
@michaelsauter @renedupont I added a question regarding the `registry` below. Agree that `secret` is right at the project level and needs to point to the x2 ods user. Regarding the additional `enabled` entry, I would initialise it to `false` at the cluster level and to `true` at the project level. Only if both (`&&`ed) evaluate to true, an Aqua scan will happen. Therefore, yes, the cluster-level means an override.

Reply:
@renedupont @michaelsauter I was also thinking of centralising the `url` property from the project-level ConfigMap inside the cluster-wide ConfigMap.

Reply:
> Regarding the additional enabled entry, I would initialise it to false at the cluster level and to true at the project level. Only if both (&&ed) evaluate to true, an Aqua scan will happen. Therefore, yes, the cluster-level means an override.

Agree!

> I was also thinking of centralising the url property from the project-level ConfigMap inside the cluster-wide ConfigMap

Would work for me.

Reply:
@michaelsauter @renedupont let's see if we can clarify the remaining question about the registry. I am not sure what this value indicates and if it should be at the project level or centralised. Is it something we can determine at ods configuration time? Ideally, we can make it a configuration parameter the users at the project-level don't have to think about.

Reply:
The docs say:

> `registry`: Refers to a name for the image registry given in the Aqua platform by an Aqua platform admin.

Not sure if this is a central registry? To me it sounds like one in which case central CM would work.

Reply:
@michaelsauter yes, I have read the docs, but assumed there was a specific reason for Rene to have the parameter defined at the project level. Not sure if Aqua has separate registries per project and if a cluster-wide central value would be sufficient? @maksymmaksym

Reply:
Hello! As there was no differentiation in the implemenation between project and global level yet before, every variable that is needed for the scannercli was in the (project-level) ConfigMap. As discussed above, in case there is only one aqua url and and one registry, it makes sense to already set it in the 'global' ConfigMap.

Reply:
Hi @renedupont ! Yes, we already clarified internally. Cheers!